core/shared/src/main/scala/scalaz/zio/stream/ZStream.scala

```diff
@@ -520,33 +520,36 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
     new ZStream[R1, E1, C] {
       override def fold[R2 <: R1, E2 >: E1, C1 >: C, S2]: Fold[R2, E2, C1, S2] =
         IO.succeedLazy { (s2, cont, f) =>
-          def feed(s1: sink.State, s2: S2, a: Chunk[A1]): ZIO[R2, E2, ZSink.Step[(sink.State, S2), A1]] =
+          def feed(s1: sink.State, s2: S2, a: Chunk[A1]): ZIO[R2, E2, (sink.State, S2, Boolean)] =
             sink.stepChunk(s1, a).flatMap { step =>
-              if (ZSink.Step.cont(step)) IO.succeed(ZSink.Step.leftMap(step)((_, s2)))
-              else {
+              if (ZSink.Step.cont(step)) {
+                IO.succeed((ZSink.Step.state(step), s2, true))
```

Comment:
This will return `extractNeeded = true` to the fold when the sink is saying that it needs more data. Should this be `false` here?

Reply:
Also, if you could add a test that uses a sink that requires more than one element until it finishes, it'd be great.

Reply:
The understanding I reached is that `feed` has three cases:

1. Sink says done, no remainder: emit downstream and exit back to the upstream fold (`f0`)
2. Sink says done, with remainder: emit downstream and loop `feed` again with the remainder
3. Sink says more: exit back to the upstream fold with the new state

In cases 1 and 2, `feed` itself does the `extract` and call to `f` to emit. In case 3, you're right that we don't want to call `extract`, _except_ in the case where the upstream runs dry, at which point you have no option but to `extract` and emit what you have, which is what the `extract.flatMap(f…)` at the end of the method seems to be for. So I guess `extractNeeded = true` means "do a final extract and emit if the input runs out".

I can add the multiple element test 👍

Reply:
@quelgar - sorry, I must be misunderstanding. This branch says that the sink requires more data. `true` is returned from `feed` which causes the `extract` to happen on line 551. Could you explain why this is correct?

Reply:
Oh and I agree with your analysis in the comment, it just doesn't seem to be reflected in the code.

Reply:
I believe it's correct because the `extract` at 551 only happens after the `self.fold` has completed, meaning the upstream cannot provide any more elements to us (right?). So the situation is the sink has asked for more data, but there isn't any more to give. In that case line 551 just extracts whatever we have (which might well fail for certain sink implementations).

In the more typical case, where the sink asks for more and the upstream _can_ provide more, then the `self.fold` keeps looping, line 551 isn't reached, and the boolean returned by `feed` is ignored completely.

Reply:
Yes, entirely true! Sorry for this dragging along. This is what happens when reviewing without checking out the code locally.